File: bindings/node/test/close-month.js
/**
 * Month-end close orchestration tests for @stateset/embedded Node.js bindings.
 *
 * generalLedger.closeMonth(periodId, options) runs depreciation, revenue
 * recognition, FX revaluation, and the period close in order; `dryRun`
 * computes candidates without writing anything.
 */

const { Commerce } = require('../index.js');
const assert = require('node:assert/strict');
const { test } = require('node:test');

async function setup(commerce) {
  const gl = commerce.generalLedger;
  await gl.initializeChartOfAccounts();

  // Wide open period covering today: GL auto-posting stamps entries with
  // today's date, which must fall inside an open period.
  const period = await gl.createPeriod({
    periodName: 'FY-wide',
    fiscalYear: 2026,
    periodNumber: 1,
    startDate: '2020-01-01',
    endDate: '2030-12-31',
  });
  assert.equal(period.status, 'future');
  const opened = await gl.openPeriod(period.id);
  assert.equal(opened.status, 'open');

  // Asset: $1200 over 12 months straight-line; all periods due by period end.
  const asset = await commerce.fixedAssets.create({
    name: 'Espresso machine',
    category: 'machinery',
    acquisitionDate: '2026-01-01',
    acquisitionCost: '1200.00',
    salvageValue: '0',
    usefulLifeMonths: 12,
    depreciationMethod: 'straight_line',
  });
  await commerce.fixedAssets.placeInService(asset.id, '2026-01-01');
  await commerce.fixedAssets.generateSchedule(asset.id);
  return { period, asset };
}

test('GeneralLedger.closeMonth', async (t) => {
  await t.test('dry run reports candidates without writing', async () => {
    const commerce = new Commerce(':memory:');
    const { period, asset } = await setup(commerce);

    const report = await commerce.generalLedger.closeMonth(period.id, { dryRun: true });
    assert.equal(report.dryRun, true);
    assert.equal(report.periodId, period.id);
    assert.equal(report.periodName, 'FY-wide');
    assert.equal(report.depreciation.status, 'dry_run');
    assert.equal(report.depreciation.entryCount, 12);
    assert.equal(report.depreciation.totalAmount, '1200.00');
    assert.equal(report.revenueRecognition.status, 'dry_run');
    assert.equal(report.fxRevaluation.status, 'skipped');
    assert.equal(report.periodClose.status, 'dry_run');
    assert.ok(report.closingEntry === null || report.closingEntry === undefined);
    assert.equal(report.periodStatus, 'open');

    // Nothing was written.
    const after = await commerce.fixedAssets.get(asset.id);
    assert.equal(after.accumulatedDepreciation, '0');
  });

  await t.test('skip flags mark steps skipped', async () => {
    const commerce = new Commerce(':memory:');
    const { period } = await setup(commerce);

    const report = await commerce.generalLedger.closeMonth(period.id, {
      skipDepreciation: true,
      skipRevenueRecognition: true,
      skipFxRevaluation: true,
      skipPeriodClose: true,
    });
    assert.equal(report.depreciation.status, 'skipped');
    assert.equal(report.revenueRecognition.status, 'skipped');
    assert.equal(report.fxRevaluation.status, 'skipped');
    assert.equal(report.periodClose.status, 'skipped');
    assert.equal(report.periodStatus, 'open');
  });

  await t.test('real run posts depreciation (period close skipped)', async () => {
    const commerce = new Commerce(':memory:');
    const { period, asset } = await setup(commerce);

    // Without GL auto-posting there is no P&L activity, so skip the final
    // period close (which requires net income) and exercise the wet path
    // for the depreciation step.
    const report = await commerce.generalLedger.closeMonth(period.id, {
      closedBy: 'binding-test',
      skipPeriodClose: true,
    });
    assert.equal(report.dryRun, false);
    assert.equal(report.depreciation.status, 'executed');
    assert.equal(report.depreciation.entryCount, 12);
    assert.equal(report.depreciation.totalAmount, '1200.00');
    assert.deepEqual(report.depreciation.warnings, []);
    assert.equal(report.fxRevaluation.status, 'skipped');
    assert.equal(report.periodClose.status, 'skipped');
    assert.equal(report.periodStatus, 'open');

    const after = await commerce.fixedAssets.get(asset.id);
    assert.equal(after.accumulatedDepreciation, '1200.00');
    assert.equal(after.status, 'fully_depreciated');
  });

  await t.test('rejects an invalid period id', async () => {
    const commerce = new Commerce(':memory:');
    await assert.rejects(
      () => commerce.generalLedger.closeMonth('not-a-uuid'),
      /Invalid UUID/,
    );
  });
});

File: crates/stateset-core/src/models/close_month.rs
//! Month-end close orchestration report types
//!
//! The "close the month" operation runs, in order: scheduled fixed-asset
//! depreciation, revenue recognition through period end, FX revaluation as of
//! period end, and the period close itself (closing entries + close period).
//! These types describe the options controlling the run and the per-step
//! report it returns. Per-item failures (e.g. a single asset that cannot post)
//! are collected as warnings and never abort the close.

use rust_decimal::Decimal;
use serde::{Deserialize, Serialize};
use std::fmt;
use uuid::Uuid;

use super::general_ledger::{JournalEntry, PeriodStatus};

/// Options controlling a month-end close run.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct CloseMonthOptions {
    /// Compute what WOULD happen (counts and amounts per step) without
    /// writing anything.
    pub dry_run: bool,
    /// Skip posting scheduled fixed-asset depreciation.
    pub skip_depreciation: bool,
    /// Skip recognizing deferred revenue through period end.
    pub skip_revenue_recognition: bool,
    /// Skip FX revaluation of foreign-currency accounts.
    pub skip_fx_revaluation: bool,
    /// Skip the final period close (closing entries + close period).
    pub skip_period_close: bool,
    /// Actor recorded as the closer. Defaults to `system`.
    pub closed_by: Option<String>,
}

/// Outcome of one step in a month-end close run.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
#[non_exhaustive]
pub enum CloseMonthStepStatus {
    /// The step ran and wrote its entries.
    Executed,
    /// The step was skipped (by flag, missing capability, or nothing to do
    /// that could be attempted, e.g. no FX account configured).
    Skipped,
    /// Candidates were computed but nothing was written (dry run).
    DryRun,
}

impl fmt::Display for CloseMonthStepStatus {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Executed => write!(f, "executed"),
            Self::Skipped => write!(f, "skipped"),
            Self::DryRun => write!(f, "dry_run"),
        }
    }
}

/// Per-step detail of a month-end close run.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CloseMonthStepReport {
    /// Whether the step executed, was skipped, or ran in dry-run mode.
    pub status: CloseMonthStepStatus,
    /// Number of entries posted (or that would be posted in a dry run):
    /// depreciation schedule entries, revenue schedule entries, revaluation
    /// journal entries, or closing entries.
    pub entry_count: u64,
    /// Total amount across those entries (depreciation posted, revenue
    /// recognized, net unrealized FX gain/loss, or closing entry debits).
    pub total_amount: Decimal,
    /// Per-item failures and notes that did not abort the close.
    pub warnings: Vec<String>,
}

impl CloseMonthStepReport {
    /// A step that was skipped, optionally with a note explaining why.
    #[must_use]
    pub fn skipped(warning: Option<String>) -> Self {
        Self {
            status: CloseMonthStepStatus::Skipped,
            entry_count: 0,
            total_amount: Decimal::ZERO,
            warnings: warning.into_iter().collect(),
        }
    }
}

/// Report returned by the month-end close orchestration.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CloseMonthReport {
    /// Period that was closed (or evaluated in dry-run mode).
    pub period_id: Uuid,
    /// Denormalized period name.
    pub period_name: String,
    /// Whether this was a dry run (nothing was written).
    pub dry_run: bool,
    /// Step 1: scheduled fixed-asset depreciation due through period end.
    pub depreciation: CloseMonthStepReport,
    /// Step 2: deferred revenue recognized through period end.
    pub revenue_recognition: CloseMonthStepReport,
    /// Step 3: FX revaluation of foreign-currency accounts as of period end.
    pub fx_revaluation: CloseMonthStepReport,
    /// Step 4: closing entries + close period.
    pub period_close: CloseMonthStepReport,
    /// The posted closing entry; `None` for dry runs, skipped closes, or a
    /// period with nothing to close.
    pub closing_entry: Option<JournalEntry>,
    /// Period status after the run (`Closed` after a real close).
    pub period_status: PeriodStatus,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn step_status_serializes_snake_case() {
        assert_eq!(
            serde_json::to_string(&CloseMonthStepStatus::DryRun).expect("serialize"),
            "\"dry_run\""
        );
        assert_eq!(CloseMonthStepStatus::Executed.to_string(), "executed");
    }

    #[test]
    fn skipped_helper_collects_warning() {
        let step = CloseMonthStepReport::skipped(Some("why".into()));
        assert_eq!(step.status, CloseMonthStepStatus::Skipped);
        assert_eq!(step.entry_count, 0);
        assert!(step.total_amount.is_zero());
        assert_eq!(step.warnings, vec!["why".to_string()]);
        assert!(CloseMonthStepReport::skipped(None).warnings.is_empty());
    }

    #[test]
    fn options_default_is_full_wet_run() {
        let options = CloseMonthOptions::default();
        assert!(!options.dry_run);
        assert!(!options.skip_depreciation);
        assert!(!options.skip_revenue_recognition);
        assert!(!options.skip_fx_revaluation);
        assert!(!options.skip_period_close);
        assert!(options.closed_by.is_none());
    }
}
